### ISSUE TITLE
Bulb menu should open immediately after clicking on the info icon in the problems view

### ISSUE THREAD

**Issue by bsutton** [microsoft/vscode#98618]
When working through large no.s of lints in the 'problems' pane I use the quick fix bulb to rapidly fix many trivial errors.

This process is slowed down by how the 'quick fix bulb' appears and then disappears.
The process I used to apply these fixes is:
Open the 'problems pane'.
click on the line
wait for the bulb to appear
click the bulb
select the action.

The problem is that the bulb will appear and then disappear before re-appearing.
It also changes color from white to yellow (which I'm sure indicates something but I've not worked it out as yet).
My preferred operation would be:
click line
click 'info' icon.
The code then looks for quick fixes and displays the menu.

This would eliminate the slight delay in having to wait for the bulb to appear.

![May-26-2020 14-38-21](https://user-images.githubusercontent.com/1078012/82907572-af98f280-9f5e-11ea-9fe5-9b038493f965.gif)

**Comment by sandy081** [comment]
> My preferred operation would be:
click line
click 'info' icon.
The code then looks for quick fixes and displays the menu.

I think this works. Did you try clicking on line and wait until lightbulb is available?

**Comment by bsutton** [comment]
Its the waiting that is the problem.
If you are working through a large no of lint errors the waiting really
slows things down.

Just let me Click the info icon and if there are quick fixes show them.
Don't make me wait for the bulb to appear.

**Comment by sandy081** [comment]
I do not think delaying the context menu on click is not a good user interaction model. Also clicking on icon to show the quick fixes might be less discoverable. We have a command that you can run to show quick fixes for the selected item in problems view. 

![image](https://user-images.githubusercontent.com/10746682/83494983-5f1e1980-a4b7-11ea-97bd-34ef1c33e7a1.png)

**Comment by bsutton** [comment]
In not suggesting that you delay the context menu.

Rather that if I click the info icon it immediately displays the context
menu.

You can still display the bulb when you are ready, just don't make me wait
for it.

**Comment by bsutton** [comment]
The short cut you mentioned is also not a particularly good solution.
I have to click on a problem line
Type the short cut
arrow down.
hit enter.

The above isn't too bad, but after I've activated the quick fix the problem
area loses focus so I'm back to using the mouse to reset the focus.

So it's still looks like my suggestion of simply changing the 'info' icon
so that clicking on it displays the context menu is the best option.

It doesn't change discoverability (you can still show the bulb when you
want to) and I can activate a quick fix with two quick clicks.
One of the 'info' icon.
One of the desired quick fix.

This workflow will allow a user to rapidly work through large no.s of
problems which is really important whenever you increase the lint level of
a project.

**Comment by DanTup** [comment]
I think an issue here might be that VS Code doesn't know a problem has quick-fixes unless it sends an async request? So you click on the problem, then there's an async request, then VS Code knows there are fixes and shows a lightbulb, which you need to click again?

If so, would a reasonable fix be for an extension to be able to set a flag on a `Diagnostic` that there are fixes and VS code can show the lightbulb in the problems view in advance? (there would still be a delay in it appearing when you click it, but it's better to `click+wait for menu` than `click+wait for bulb+click+wait for menu`).

In Dart, the server already computes this info. We have a `hasFix` boolean on all errors, but it's not mappable to anything in VS Code.

**Comment by sandy081** [comment]
> Rather that if I click the info icon it immediately displays the context menu.

Yes we do not know upfront if we have quick fixes upfront and needs to ask extension to provide them. Hence this delay is unavoidable.

I do not want to drag this request into a separate discussion asking for API change. If API change is needed for this, I do not think it is worth considering.

**Comment by DanTup** [comment]
It might be possible without an API change, eg. by using `DiagnosticTag`? I think it would be a fairly nice improvement if VS code knew which problems had quick-fixes from the problems view. Imagine something like enabling a new lint which generates a lot of issues - it would be much nicer to work through them if the bulb was visible/clickable in the list.

**Comment by sandy081** [comment]
cc @mjbvz if you want to consider this request as you own code actions and from problems perspective I would want to align with the editor code actions.

**Comment by bsutton** [comment]
Sounds like this can be done without an API change.

Click info icon.
Async call
Display menu

If no quick fixes are available display a single item
'No quick fixes available.'

**Comment by bsutton** [comment]
> cc @mjbvz if you want to consider this request as you own code actions and from problems perspective I would want to align with the editor code actions.

I don't understand this statement?

**Comment by sandy081** [comment]
I do not think this is a good user interaction to show context menu after some delay.

**Comment by bsutton** [comment]
The delay should be fairly minor.
You could change the info icon to a spinning gif until the context menu
appears.

**Comment by sandy081** [comment]
> The delay should be fairly minor.

Depends and we cannot rely on that. 

Well, I see the request but I do not think it meet the priorities. So I have to tag this as out of scope and we would revisit this when there is a demand.

**Comment by DanTup** [comment]
@sandy081 any reason why this couldn't be done using a DiagnosticTag or something as suggested above? It removes any delay between clicking/menu opening, and would be a nice usability improvement showing which problems have quickfixes in the problem list without having to click (the current behaviour already feels kinda awkward).

**Comment by sandy081** [comment]
> any reason why this couldn't be done using a DiagnosticTag or something as suggested above?

Can be done but not in priorities as mentioned above.

**Comment by bsutton** [comment]
I can't say that I'm happy with the concept of closing issues out.

Even if you aren't going to do this now, having an open ticket lets other users find and vote on this ticket.

**Comment by bsutton** [comment]
I also disagree with the priority position on this one.
Quick fixes are one of the most common elements that people interact with with in the UI.
Whilst it is a minor announce it is a high use work flow and as such I would suggest that it needs greater priority.
It falls in to the category which is often referred to as a paper cut. Its small and hurts a little bit, but when you are covered in them its a real pain.

**Comment by sandy081** [comment]
> Even if you aren't going to do this now, having an open ticket lets other users find and vote on this ticket.

Sorry for closing this. Reopened it and moved to backlog candidates.

**Comment by bsutton** [comment]
It's still showing up as closed.

**Comment by DanTup** [comment]
A similar request came up at https://github.com/Dart-Code/Dart-Code/issues/3254. The request was to have a single-click icon that would apply the fix. Fixes can already by marked as "preferred" fixes, so if they're populated, it would be nice if you could just work through the problems list clicking the fix buttons without even needing to open the menu (and perform two clicks).

Like this issue, I think it would require fetching the quick-fixes prior to the user clicking on the problem - just since the problems view only generally shows a small number of problems, maybe it would be possible to fetch a few eagerly?

If you had a large number of errors from enabling a new lint (or upgrading an SDK that generates lots of deprecation warnings with quick-fixes), you could get through them a _lot_ faster that way.

**Comment by devgid** [comment]
@bsutton since your issue is fixing a large number of lint problems, then would a github issue like [this ](https://github.com/microsoft/vscode/issues/209212)not be better solution? Instead of clicking on item in the problem list each time, it gives the context menu of the next item as soon as the previous item selection if fixed?

**Comment by bsutton** [comment]
@devgid the other solution isn't terrible, but I would argue that my solution solves other problems as well - even for a single fix the lag is annoying.
